cluster/job_once_scheduler.go

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package cluster
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// syncedCallback uses the mutex to make things predictable for the client: the callback will be
+// called once at a time (the client does not need to worry about concurrency within the callback)
+type syncedCallback struct {
+	mu       sync.Mutex
+	callback func(string)
+}
+
+type syncedJobs struct {
+	mu   sync.RWMutex
+	jobs map[string]*JobOnce
+}
+
+type JobOnceScheduler struct {
+	pluginAPI JobPluginAPI
+
+	startedMu sync.RWMutex
+	started   bool
+
+	activeJobs     *syncedJobs
+	storedCallback *syncedCallback
+}
+
+var schedulerOnce sync.Once
+var s *JobOnceScheduler
+
+// GetJobOnceScheduler returns a scheduler which is ready to accept callbacks. Repeated calls will
+// return the same scheduler.
+func GetJobOnceScheduler(pluginAPI JobPluginAPI) *JobOnceScheduler {
```

Comment:
Not sure what you mean by "instance" (I could think of a couple different things you might mean, so instead of guessing I'll let you clarify then I can answer :) )

Reply:
I mean just having the plugin create an instance of the scheduler rather than having this function that creates a singleton.

Reply:
In that case: we don't want the plugin to be able to create more than one instance of the scheduler -- there should only be one poller, and one set of jobs running per server.

Reply:
Ok. I generally try to avoid singletons as I find them to be mostly unnecessary. 1/5 though.

Reply:
Ohhhhhhh I see. 👍

Reply:
Haven't reviewed the code here, but agree with @crspeller -- while we should discourage creating more than one scheduler, I don't think we should outright disallow it -- it should work just as if two servers were running within a single instance. I also find it often results in having to jump through hoops when unit testing.

Reply:
@lieut-data Normally I would agree with you both, but in this case I think a singleton makes sense: because we're making sure to start up all Y jobs currently scheduled in the db.
In other words, if we allow the user to create X schedulers, we would have X copies of the same set of Y jobs running on the same server.
To prevent this duplication, we could make each server have a map of currently scheduled jobs, but that's a singleton too (and kind of confusing if we're allowing any number of schedulers).

Reply:
We are not saying that there should be more than one, just that it's unnecessary to actually restrict it. Also yes it can interfere with unit testing efforts. 
(also my :+1: comment was meant for the other thread)

Reply:
Hmm, maybe I'm misunderstanding what you mean then. What do you mean `it's unnecessary to actually restrict it`? Do you mean, allow the user to create new schedulers each time `GetJobOnceScheduler` is called?